lib/Animator/Animator.man
'\" t
.TH "" 3 "" "Version 3.0" "Free Widget Foundation"
.SH NAME
XfwfAnimator
.SH DESCRIPTION
The animator widget cycles through a series of images (XImage
structures). Each image is displayed for a certain number of
milliseconds. After the last image, the animation can either start
over or stop with the last image displayed.

The animation must be started with \fIXfwfStartAnimation\fP and it may be
stopped with \fIXfwfStopAnimation\fP.

The widget uses the location resources (see XfwfBoard(3)) for
positioning only; the preferred size of the widget will be the size of
the largest image, plus room for the frame.

.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfAnimator
Name	Class	Type	Default
XtNimages	XtCImages	ImageList	NULL 
XtNintervals	XtCIntervals	CardinalList	NULL 
XtNdefaultInterval	XtCDefaultInterval	Cardinal 	500 
XtNcycle	XtCCycle	Boolean 	True 

.TE
.ps +2

.TP
.I "XtNimages"
The list of images can be specified in the resource file as a series
of filenames of files in XPM format. If it is set from an application,
the list must end with a \fINULL\fP. Every element is a pointer to an
\fIXImage\fP structure.

.TP
.I "XtNintervals"
The time that each image will be displayed can be set with
\fIintervals\fP. It is a list of cardinals, each representing a number of
milliseconds. The list must end with a zero. If there are insufficient
numbers in the list, the last number will be used as the interval for
any remaining images. If the list is empty (NULL), the interval will
be \fIdefaultInterval\fP (normally 500).

.TP
.I "XtNdefaultInterval"
If the list of intervals is empty, the interval will be
\fIdefaultInterval\fP milliseconds.

.TP
.I "XtNcycle"
The animation repeats indefinitely by default. When the \fIcycle\fP
resource is set to \fIFalse\fP the animation sequence is shown once only,
for every call to \fIXfwfStartAnimation\fP.

.TP
.I "XtNframeType"
The default type of frame is set to `sunken'.

.TP
.I "XtNrel_width"
If there is no image to display, the widget has no obvious size. The
following defaults for the location resources create a 7 by 7 pixel
widget.

.TP
.I "XtNabs_width"

.TP
.I "XtNrel_height"

.TP
.I "XtNabs_height"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfBoard
Name	Class	Type	Default
XtNabs_x	XtCAbs_x	Position 	0 
XtNrel_x	XtCRel_x	Float 	"0.0"
XtNabs_y	XtCAbs_y	Position 	0 
XtNrel_y	XtCRel_y	Float 	"0.0"
XtNabs_width	XtCAbs_width	Position 	0 
XtNrel_width	XtCRel_width	Float 	"1.0"
XtNabs_height	XtCAbs_height	Position 	0 
XtNrel_height	XtCRel_height	Float 	"1.0"
XtNhunit	XtCHunit	Float 	"1.0"
XtNvunit	XtCVunit	Float 	"1.0"
XtNlocation	XtCLocation	String 	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfFrame
Name	Class	Type	Default
XtNcursor	XtCCursor	Cursor 	None 
XtNframeType	XtCFrameType	FrameType 	XfwfRaised 
XtNframeWidth	XtCFrameWidth	Dimension 	0 
XtNouterOffset	XtCOuterOffset	Dimension 	0 
XtNinnerOffset	XtCInnerOffset	Dimension 	0 
XtNshadowScheme	XtCShadowScheme	ShadowScheme 	XfwfAuto 
XtNtopShadowColor	XtCTopShadowColor	Color 	compute_topcolor 
XtNbottomShadowColor	XtCBottomShadowColor	Color 	compute_bottomcolor 
XtNtopShadowStipple	XtCTopShadowStipple	Bitmap 	NULL 
XtNbottomShadowStipple	XtCBottomShadowStipple	Bitmap 	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfCommon
Name	Class	Type	Default
XtNuseXCC	XtCUseXCC	Boolean 	TRUE 
XtNusePrivateColormap	XtCUsePrivateColormap	Boolean 	FALSE 
XtNuseStandardColormaps	XtCUseStandardColormaps	Boolean 	TRUE 
XtNstandardColormap	XtCStandardColormap	Atom 	0 
XtNxcc	XtCXCc	XCC 	create_xcc 
XtNtraversalOn	XtCTraversalOn	Boolean 	True 
XtNhighlightThickness	XtCHighlightThickness	Dimension 	2 
XtNhighlightColor	XtCHighlightColor	Color 	XtDefaultForeground 
XtNbackground	XtCBackground	Color 	XtDefaultBackground 
XtNhighlightPixmap	XtCHighlightPixmap	Pixmap 	None 
XtNnextTop	XtCNextTop	Callback	NULL 
XtNuserData	XtCUserData	Pointer	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Composite
Name	Class	Type	Default
XtNchildren	XtCChildren	WidgetList 	NULL 
insertPosition	XtCInsertPosition	XTOrderProc 	NULL 
numChildren	XtCNumChildren	Cardinal 	0 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2

.SS "Exports"

To start the animation an application should call
\fIXfwfStartAnimation\fP. If the animation is already running, it will be
restarted from the beginning.

.nf
XfwfStartAnimation( $)
.fi

The animation can be stopped with \fIXfwfStopAnimation\fP.

.nf
XfwfStopAnimation( $)
.fi



debian/changelog
fwf (4.0.1-1) unstable; urgency=low

  * Build using autoconf/automake.
  * Compile for x86 and amd64 processors.

 -- Erling Andersen <Erling.Andersen@Helse-Bergen.NO>  Mon, 09 Feb 2009 10:23:59 +0100


lib/Animator/Animator.c
/* Generated by wbuild from "Animator.w"
** (generator version $Revision$ of $Date$)
*/
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 365 "Animator.w"
#include <X11/xpm.h>
#line 366 "Animator.w"
#include <stdio.h>
#include <Xfwf/AnimatorP.h>
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 110 "Animator.w"
static void class_initialize(
#if NeedFunctionPrototypes
void
#endif
);
#line 122 "Animator.w"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 149 "Animator.w"
static void start_animation(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 169 "Animator.w"
static void stop_animation(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 181 "Animator.w"
static void expose(
#if NeedFunctionPrototypes
Widget,XEvent *,Region 
#endif
);
#line 199 "Animator.w"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 234 "Animator.w"
static void next_image(
#if NeedFunctionPrototypes
XtPointer ,XtIntervalId *
#endif
);
#line 268 "Animator.w"
#define done(type, value) do {\
      if (to->addr != NULL) {\
	  if (to->size < sizeof(type)) {\
	      to->size = sizeof(type);\
	      return False;\
	  }\
	  *(type*)(to->addr) = (value);\
      } else {\
	  static type static_val;\
	  static_val = (value);\
	  to->addr = (XtPointer)&static_val;\
      }\
      to->size = sizeof(type);\
      return True;\
  }while (0 )


#line 285 "Animator.w"
static Boolean  cvtStringToImageList(
#if NeedFunctionPrototypes
Display *,XrmValuePtr ,Cardinal *,XrmValuePtr ,XrmValuePtr ,XtPointer *
#endif
);
#line 337 "Animator.w"
static Boolean  cvtStringToCardinalList(
#if NeedFunctionPrototypes
Display *,XrmValuePtr ,Cardinal *,XrmValuePtr ,XrmValuePtr ,XtPointer *
#endif
);
#line 234 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 234 "Animator.w"
static void next_image(XtPointer  client_data,XtIntervalId * timer)
#else
#line 234 "Animator.w"
static void next_image(client_data,timer)XtPointer  client_data;XtIntervalId * timer;
#endif
#line 235 "Animator.w"
{
    Widget self = (Widget) client_data;
    Cardinal delay;

    if (((XfwfAnimatorWidget)self)->xfwfAnimator.images[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image+1] == NULL) {
	if (! ((XfwfAnimatorWidget)self)->xfwfAnimator.cycle) return;			/* Nothing more */
	((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image = 0;
	((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval = 0;
	if (((XfwfAnimatorWidget)self)->xfwfAnimator.intervals == NULL || ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[0] == 0)
	    delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.defaultInterval;
	else
	    delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[0];
    } else {
	((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image++;
	if (((XfwfAnimatorWidget)self)->xfwfAnimator.intervals == NULL)
	    delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.defaultInterval;
	else if (((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval+1] == 0)
	    delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval];
	else {
	    ((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval++;
	    delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval];
	}
    }
    ((XfwfAnimatorWidgetClass)self->core.widget_class)->core_class.expose(self, NULL, NULL);
    ((XfwfAnimatorWidget)self)->xfwfAnimator.timer = XtAppAddTimeOut(XtWidgetToApplicationContext(self), delay,
			     next_image, self);
}
#line 285 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 285 "Animator.w"
static Boolean  cvtStringToImageList(Display * display,XrmValuePtr  args,Cardinal * num_args,XrmValuePtr  from,XrmValuePtr  to,XtPointer * converter_data)
#else
#line 285 "Animator.w"
static Boolean  cvtStringToImageList(display,args,num_args,from,to,converter_data)Display * display;XrmValuePtr  args;Cardinal * num_args;XrmValuePtr  from;XrmValuePtr  to;XtPointer * converter_data;
#endif
#line 286 "Animator.w"
{
    String p;
    XImage *im, *shape, **list = NULL;
    int status, n = 0;
    Cardinal one = 1;

    if (*num_args != 0)
	XtAppErrorMsg
	    (XtDisplayToApplicationContext(display),
	     "cvtStringToImageList", "wrongParameters",
	     "XtToolkitError",
	     "String to image list type conversion needs no arguments", 
	     (String*) NULL, (Cardinal*) NULL);

    p = strtok((String) from->addr, " \t\n,");
    while (p) {
	status = XpmReadFileToImage(display, p, &im, &shape, NULL);
	switch (status) {
	case XpmOpenFailed:
	case XpmFileInvalid:
	case XpmNoMemory:
	    XtAppWarningMsg
		(XtDisplayToApplicationContext(display),
		 "cvtStringToImageList", "fileError",
		 "XtToolkitError",
		 "Failed to read image from \"%s\"",
		 &p, &one);
	    break;	    
	case XpmColorError:
	case XpmColorFailed:
	    XtAppWarningMsg
		(XtDisplayToApplicationContext(display),
		 "cvtStringToImageList", "allocColor",
		 "XtToolkitError",
		 "Could not get (all) colors for image \"%s\"",
		 &p, &one);
	case XpmSuccess:
	    n++;
	    list = (XImage**) XtRealloc((char*) list, n * sizeof(im));
	    list[n-1] = im;
	}
	p = strtok(NULL, " \t\n,");
    }
    if (n > 0) {
	n++;
	list = (XImage**) XtRealloc((char*) list, n * sizeof(im));
	list[n-1] = NULL;
    }
    done(XImage**, list);
}
#line 337 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 337 "Animator.w"
static Boolean  cvtStringToCardinalList(Display * display,XrmValuePtr  args,Cardinal * num_args,XrmValuePtr  from,XrmValuePtr  to,XtPointer * converter_data)
#else
#line 337 "Animator.w"
static Boolean  cvtStringToCardinalList(display,args,num_args,from,to,converter_data)Display * display;XrmValuePtr  args;Cardinal * num_args;XrmValuePtr  from;XrmValuePtr  to;XtPointer * converter_data;
#endif
#line 338 "Animator.w"
{
    String p;
    Cardinal *list = NULL;
    int n = 0, h, stat;

    if (*num_args != 0)
	XtAppErrorMsg
	    (XtDisplayToApplicationContext(display),
	     "cvtStringToCardinalList", "wrongParameters",
	     "XtToolkitError",
	     "String to cardinal list type conversion needs no arguments", 
	     (String*) NULL, (Cardinal*) NULL);

    p = (char *) from->addr;
    do {
	n++;
	list = (Cardinal*) XtRealloc((char*) list, n * sizeof(Cardinal));
	/* list[n-1] = strtoul(p, &p, 0); */
	stat = sscanf(p, "%u%n", &list[n-1], &h);
	p = p + h;
    } while (stat == 1 && list[n-1] != 0);
    list[n-1] = 0;
    done(Cardinal *, list);
}

static XtResource resources[] = {
#line 27 "Animator.w"
{XtNimages,XtCImages,XtRImageList,sizeof(((XfwfAnimatorRec*)NULL)->xfwfAnimator.images),XtOffsetOf(XfwfAnimatorRec,xfwfAnimator.images),XtRImmediate,(XtPointer)NULL },
#line 36 "Animator.w"
{XtNintervals,XtCIntervals,XtRCardinalList,sizeof(((XfwfAnimatorRec*)NULL)->xfwfAnimator.intervals),XtOffsetOf(XfwfAnimatorRec,xfwfAnimator.intervals),XtRImmediate,(XtPointer)NULL },
#line 41 "Animator.w"
{XtNdefaultInterval,XtCDefaultInterval,XtRCardinal,sizeof(((XfwfAnimatorRec*)NULL)->xfwfAnimator.defaultInterval),XtOffsetOf(XfwfAnimatorRec,xfwfAnimator.defaultInterval),XtRImmediate,(XtPointer)500 },
#line 47 "Animator.w"
{XtNcycle,XtCCycle,XtRBoolean,sizeof(((XfwfAnimatorRec*)NULL)->xfwfAnimator.cycle),XtOffsetOf(XfwfAnimatorRec,xfwfAnimator.cycle),XtRImmediate,(XtPointer)True },
#line 51 "Animator.w"
{XtNframeType,XtCFrameType,XtRFrameType,sizeof(((XfwfAnimatorRec*)NULL)->xfwfFrame.frameType),XtOffsetOf(XfwfAnimatorRec,xfwfFrame.frameType),XtRImmediate,(XtPointer)XfwfSunken },
#line 57 "Animator.w"
{XtNrel_width,XtCRel_width,XtRFloat,sizeof(((XfwfAnimatorRec*)NULL)->xfwfBoard.rel_width),XtOffsetOf(XfwfAnimatorRec,xfwfBoard.rel_width),XtRString,(XtPointer)"0.0"},
#line 58 "Animator.w"
{XtNabs_width,XtCAbs_width,XtRPosition,sizeof(((XfwfAnimatorRec*)NULL)->xfwfBoard.abs_width),XtOffsetOf(XfwfAnimatorRec,xfwfBoard.abs_width),XtRImmediate,(XtPointer)7 },
#line 59 "Animator.w"
{XtNrel_height,XtCRel_height,XtRFloat,sizeof(((XfwfAnimatorRec*)NULL)->xfwfBoard.rel_height),XtOffsetOf(XfwfAnimatorRec,xfwfBoard.rel_height),XtRString,(XtPointer)"0.0"},
#line 60 "Animator.w"
{XtNabs_height,XtCAbs_height,XtRPosition,sizeof(((XfwfAnimatorRec*)NULL)->xfwfBoard.abs_height),XtOffsetOf(XfwfAnimatorRec,xfwfBoard.abs_height),XtRImmediate,(XtPointer)7 },
};

XfwfAnimatorClassRec xfwfAnimatorClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &xfwfBoardClassRec,
/* class_name   	*/  "XfwfAnimator",
/* widget_size  	*/  sizeof(XfwfAnimatorRec),
/* class_initialize 	*/  class_initialize,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  XtInheritRealize,
/* actions      	*/  NULL,
/* num_actions  	*/  0,
/* resources    	*/  resources,
/* num_resources 	*/  9,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  True ,
/* compress_exposure 	*/  XtExposeCompressMultiple ,
/* compress_enterleave 	*/  True ,
/* visible_interest 	*/  TRUE ,
/* destroy      	*/  NULL,
/* resize       	*/  XtInheritResize,
/* expose       	*/  expose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  NULL,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* composite_class part */
XtInheritGeometryManager,
XtInheritChangeManaged,
XtInheritInsertChild,
XtInheritDeleteChild,
NULL
},
{ /* XfwfCommon_class part */
XtInherit_compute_inside,
XtInherit_total_frame_width,
XtInherit_highlight_border,
XtInherit_unhighlight_border,
XtInherit_would_accept_focus,
XtInherit_traverse,
XtInherit_lighter_color,
XtInherit_darker_color,
/* traversal_trans */  NULL ,
},
{ /* XfwfFrame_class part */
 /* dummy */  0
},
{ /* XfwfBoard_class part */
XtInherit_set_abs_location,
},
{ /* XfwfAnimator_class part */
start_animation,
stop_animation,
},
};
WidgetClass xfwfAnimatorWidgetClass = (WidgetClass) &xfwfAnimatorClassRec;
static void _resolve_inheritance(class)
WidgetClass class;
{
  XfwfAnimatorWidgetClass c = (XfwfAnimatorWidgetClass) class;
  XfwfAnimatorWidgetClass super;
  static CompositeClassExtensionRec extension_rec = {
    NULL, NULLQUARK, XtCompositeExtensionVersion,
    sizeof(CompositeClassExtensionRec), True};
  CompositeClassExtensionRec *ext;
  ext = (XtPointer)XtMalloc(sizeof(*ext));
  *ext = extension_rec;
  ext->next_extension = c->composite_class.extension;
  c->composite_class.extension = ext;
  if (class == xfwfAnimatorWidgetClass) return;
  super = (XfwfAnimatorWidgetClass)class->core_class.superclass;
  if (c->xfwfAnimator_class.start_animation == XtInherit_start_animation)
    c->xfwfAnimator_class.start_animation = super->xfwfAnimator_class.start_animation;
  if (c->xfwfAnimator_class.stop_animation == XtInherit_stop_animation)
    c->xfwfAnimator_class.stop_animation = super->xfwfAnimator_class.stop_animation;
}
#line 110 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 110 "Animator.w"
static void class_initialize(void)
#else
#line 110 "Animator.w"
static void class_initialize()
#endif
#line 111 "Animator.w"
{
    XtSetTypeConverter(XtRString, XtRImageList, cvtStringToImageList,
		       NULL, 0, XtCacheNone, NULL);
    XtSetTypeConverter(XtRString, XtRCardinalList, cvtStringToCardinalList,
		       NULL, 0, XtCacheNone, NULL);
}
#line 122 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 122 "Animator.w"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 122 "Animator.w"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 123 "Animator.w"
{
    Dimension wd, ht, dummy1, dummy2;
    Position x, y;
    XImage **p;
    XtGCMask value_mask = 0;
    XGCValues values;

    ((XfwfAnimatorWidget)self)->xfwfAnimator.timer = NULL;
    ((XfwfAnimatorWidget)self)->xfwfAnimator.image_gc = XtGetGC(self, value_mask, &values);
    if (((XfwfAnimatorWidget)self)->xfwfAnimator.images != NULL) {
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &dummy1, &dummy2);
	wd = 1; ht = 1;
	for (p = ((XfwfAnimatorWidget)self)->xfwfAnimator.images; *p; p++) {
	    if ((*p)->width > wd) wd = (*p)->width;
	    if ((*p)->height > ht) ht = (*p)->height;
	}
	XtVaSetValues(self, XtNwidth, wd + 2*x, XtNheight, ht + 2*y, NULL);
    }
}
#line 149 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 149 "Animator.w"
static void start_animation(Widget self)
#else
#line 149 "Animator.w"
static void start_animation(self)Widget self;
#endif
#line 150 "Animator.w"
{
    Cardinal delay;

    if (! ((XfwfAnimatorWidget)self)->xfwfAnimator.images) return;
    if (((XfwfAnimatorWidget)self)->xfwfAnimator.timer) XtRemoveTimeOut(((XfwfAnimatorWidget)self)->xfwfAnimator.timer);
    ((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image = 0;
    ((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval = 0;
    if (! ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals || ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[0] == 0)
	delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.defaultInterval;
    else
	delay = ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_interval];
    ((XfwfAnimatorWidgetClass)self->core.widget_class)->core_class.expose(self, NULL, NULL);
    ((XfwfAnimatorWidget)self)->xfwfAnimator.timer = XtAppAddTimeOut(XtWidgetToApplicationContext(self), delay,
			     next_image, self);
}
#line 169 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 169 "Animator.w"
static void stop_animation(Widget self)
#else
#line 169 "Animator.w"
static void stop_animation(self)Widget self;
#endif
#line 170 "Animator.w"
{
    if (((XfwfAnimatorWidget)self)->xfwfAnimator.timer) {
	XtRemoveTimeOut(((XfwfAnimatorWidget)self)->xfwfAnimator.timer);
	((XfwfAnimatorWidget)self)->xfwfAnimator.timer = NULL;
    }
}
#line 181 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 181 "Animator.w"
static void expose(Widget self,XEvent * event,Region  region)
#else
#line 181 "Animator.w"
static void expose(self,event,region)Widget self;XEvent * event;Region  region;
#endif
#line 182 "Animator.w"
{
    Dimension wd, ht;
    Position x, y;

    if (! XtIsRealized(self)) return;
    if (region != NULL)
	XSetRegion(XtDisplay(self), ((XfwfAnimatorWidget)self)->xfwfAnimator.image_gc, region);
    ((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &wd, &ht);
    if (((XfwfAnimatorWidget)self)->xfwfAnimator.images) XPutImage(XtDisplay(self), XtWindow(self), ((XfwfAnimatorWidget)self)->xfwfAnimator.image_gc,
			   ((XfwfAnimatorWidget)self)->xfwfAnimator.images[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image], 0, 0, x, y,
			   ((XfwfAnimatorWidget)self)->xfwfAnimator.images[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image]->width,
			   ((XfwfAnimatorWidget)self)->xfwfAnimator.images[((XfwfAnimatorWidget)self)->xfwfAnimator.cur_image]->height);
    if (region != NULL)
	XSetClipMask(XtDisplay(self), ((XfwfAnimatorWidget)self)->xfwfAnimator.image_gc, None);
    xfwfBoardClassRec.core_class.expose(self, event, region);
}
#line 199 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 199 "Animator.w"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 199 "Animator.w"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 200 "Animator.w"
{
    Dimension wd, ht, dummy1, dummy2;
    Position x, y;
    XImage **p;
    Boolean restart = False;

    if (((XfwfAnimatorWidget)old)->xfwfAnimator.images != ((XfwfAnimatorWidget)self)->xfwfAnimator.images) {
	if (((XfwfAnimatorWidget)self)->xfwfAnimator.timer) restart = True;
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfAnimator_class.stop_animation(self);
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &dummy1, &dummy2);
	wd = 1; ht = 1;
	for (p = ((XfwfAnimatorWidget)self)->xfwfAnimator.images; *p; p++) {
	    if ((*p)->width > wd) wd = (*p)->width;
	    if ((*p)->height > ht) ht = (*p)->height;
	}
	XtVaSetValues(self, XtNwidth, wd + 2*x, XtNheight, ht + 2*y, NULL);
    }
    if (((XfwfAnimatorWidget)old)->xfwfAnimator.intervals != ((XfwfAnimatorWidget)self)->xfwfAnimator.intervals) {
	if (((XfwfAnimatorWidget)self)->xfwfAnimator.timer) restart = True;
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfAnimator_class.stop_animation(self);
    }
    if (restart) ((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfAnimator_class.start_animation(self);
    return False;
}
#line 87 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 87 "Animator.w"
void XfwfStartAnimation(Widget self)
#else
#line 87 "Animator.w"
void XfwfStartAnimation(self)Widget self;
#endif
#line 88 "Animator.w"
{
    if (! XtIsSubclass(self, xfwfAnimatorWidgetClass))
	XtWarning("Illegal widget type in XfwfStartAnimation");
    else
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfAnimator_class.start_animation(self);
}
#line 97 "Animator.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 97 "Animator.w"
void XfwfStopAnimation(Widget self)
#else
#line 97 "Animator.w"
void XfwfStopAnimation(self)Widget self;
#endif
#line 98 "Animator.w"
{
    if (! XtIsSubclass(self, xfwfAnimatorWidgetClass))
	XtWarning("Illegal widget type in XfwfStopAnimation");
    else
	((XfwfAnimatorWidgetClass)self->core.widget_class)->xfwfAnimator_class.stop_animation(self);
}
